mars/session.py

```diff
@@ -101,58 +105,51 @@ def run(self, *tensors, **kw):
             ret_list = True
 
         tensors = tuple(mt.tensor(t) for t in tensors)
-        results = [None] * len(tensors)
-        idx_to_run_tensors = OrderedDict()
-        idx_to_fetch_tensors = OrderedDict()
-
-        # those executed tensors should fetch data directly, submit the others
-        for i, t in enumerate(tensors):
-            if t.key in self._executed_keys:
-                idx_to_fetch_tensors[i] = t
-            else:
-                idx_to_run_tensors[i] = t
-
-        # execute the non-executed tensors
-        if idx_to_run_tensors:
-            execute_result = self._sess.run(*idx_to_run_tensors.values(), **kw)
-            if execute_result:
-                # fetch is True
-                for j, result in zip(idx_to_run_tensors, execute_result):
-                    results[j] = result
-            run_tensors = list(idx_to_run_tensors.values())
-            self._executed_keys.update(t.key for t in run_tensors)
-            for t in run_tensors:
-                t._execute_session = self
+        result = self._sess.run(*tensors, **kw)
+
+        for t in tensors:
+            t._execute_session = self
 
         for t in tensors:
             if np.nan in t.shape:
                 self._sess._update_tensor_shape(t)
 
         if fetch:
-            # do fetch
-            if idx_to_fetch_tensors:
-                for j, result in zip(idx_to_fetch_tensors,
-                                     self._sess.fetch(*idx_to_fetch_tensors.values(), **kw)):
-                    results[j] = result
-
             ret = []
-            for r, t in zip(results, tensors):
+            for r, t in zip(result, tensors):
                 if r is None:
                     ret.append(r)
                     continue
                 if t.isscalar() and hasattr(r, 'item'):
                     ret.append(np.asscalar(r))
                 else:
                     ret.append(r)
-
             if ret_list:
                 return ret
             return ret[0]
 
-    def fetch(self, tensor):
-        if tensor.key not in self._executed_keys:
-            raise ValueError('Cannot fetch the unexecuted tensor')
-        return self._sess.fetch(tensor)
+    def fetch(self, *tensors, **kw):
+        ret_list = False
+        if len(tensors) == 1 and isinstance(tensors[0], (tuple, list)):
+            ret_list = True
+            tensors = tensors[0]
+        elif len(tensors) > 1:
+            ret_list = True
+
+        result = self._sess.fetch(*tensors, **kw)
+
+        ret = []
+        for r, t in zip(result, tensors):
+            if r is None:
+                ret.append(r)
```

Comment:
Not covered

Reply:
Just copy the logic from `run`, does any tensor's return value is `None`?

Reply:
It should be removed as `run` will return None when specify `fetch=False` which won't happened in `fetch`